walker.go

```diff
@@ -0,0 +1,429 @@
+func (w *Walker) Iterate(visitor Visitor) error {
+
+	// Iterate until either: the end of the DAG (`errUpOnRoot`), a `Pause`
+	// is requested (`errPauseWalkOperation`) or an error happens (while
+	// going down).
+	for {
+
+		// First, go down as much as possible.
+		for {
+			err := w.down(visitor)
+
+			if err == ErrDownNoChild {
+				break
+				// Can't keep going down from this node, try to move Next.
+			}
+
+			if err == errPauseWalkOperation {
+				return nil
+				// Pause requested, `errPauseWalkOperation` is just an internal
+				// error to signal to pause, don't pass it along.
+			}
+
+			if err != nil {
+				return err
+				// `down` is the only movement that can return *any* error.
+			}
+		}
+
+		// Can't move down anymore, turn to the next child in the `ActiveNode`
+		// to go down a different path. If there are no more child nodes
+		// available, go back up.
+		for {
+			err := w.Next()
+			if err == nil {
+				break
+				// No error, it turned to the next child. Try to go down again.
+			}
+
+			// It can't go Next (`ErrNextNoChild`), try to move up.
+			err = w.up()
+			if err != nil {
+				// Can't move up, on the root again (`errUpOnRoot`).
+				return EndOfDag
+			}
+
+			// Moved up, try `Next` again.
+		}
+
+		// Turned to the next child (after potentially many up moves),
+		// try going down again.
+	}
+}
+
+// Search a specific node in a downwards manner. The `Visitor` should be
+// used to steer the search selecting at each node which child will the
+// search continue to (extending the `path` in that direction) or pause it
+// (if the desired node has been found). The search always starts from
+// the root. It modifies the position so it shouldn't be used in-between
+// `Iterate` calls (it can be used to set the position *before* iterating).
+// If the visitor returns any non-`nil` errors the search will stop.
+//
+// TODO: The search could be extended to search from the current position.
+// (Is there something in the logic that would prevent it at the moment?)
+func (w *Walker) Search(visitor Visitor) error {
+
+	if visitor == nil {
+		return ErrNilVisitor
+		// Although valid, there is no point in calling `Search` without
+		// any extra logic, it would just go down to the leftmost leaf,
+		// so this would probably be a user error.
+	}
+
+	// Go down until it the desired node is found (that will be signaled
+	// pausing the search with `errPauseWalkOperation`) or a leaf node is
+	// reached (end of the DAG).
+	for {
+		err := w.down(visitor)
+
+		if err == errPauseWalkOperation {
+			return nil
+			// Found the node, `errPauseWalkOperation` is just an internal
+			// error to signal to pause, don't pass it along.
+		}
+
+		if err == ErrDownNoChild {
+			return nil
+			// Can't keep going down from this node, either at a leaf node
+			// or the `Visitor` has moved the child index past the
+			// available index (probably because none indicated that the
+			// target node could be down from there).
```

Comment:
I kind of noted this in the comment above, but it seems like this should be a new type of public error if we aren’t on a leaf, right? In that case, we failed to actually find what we were looking for.

Reply:
I think it’s reasonable that explicit pausing and pausing because we reached the end are the same. (It looks to me like reaching the end here ought to be the normal situation, right? So returning an error would feel odd to me.)

I was thinking more of this case:

```go
err := walker.Seek(func(node NavigableNode) error {
    // Say this node only has one child, but our logic naively moved past it because the offsets provided in the node were bad or because we just wrote poor code
    walker.NextChild()
    return nil
})
err == nil
```

The visitor would have never seen a leaf node, but nothing here makes it obvious something went wrong.

Reply:
You need to check `NextChild` for errors.

> The visitor would have never seen a leaf node, but nothing here makes it obvious something went wrong.

The visitor would have seen (at least) the root node, which in your example is a leaf node (empty DAGs are not allowed) so I wouldn't consider it wrong, it would be like searching an array of only one value (a sad search, yes, but I wouldn't see it as an error).

Reply:
> The visitor would have seen (at least) the root node, which in your example is a leaf node (empty DAGs are not allowed)

I don’t think you’d have to have an empty DAG to hit that example, though. Wouldn’t this DAG fit the example?

```
           root
            |
  +---------+----------+
  |                    |
node 1               node 2
  |                    |
  |               +----+----+
  |               |         |
node 3          node 4    node 5
```

That said, I just realized I misread something: I thought `NextChild()` returned an `ErrDownNoChild`, which would get swallowed by `Seek()`, so this wouldn’t do any good:

```go
err := walker.Seek(func(node NavigableNode) error {
    // Say this node only has one child, but our logic naively moved
    // past it because the offsets provided in the node were bad or
    // because we just wrote poor code
    return walker.NextChild()
})
err == nil
```

but it returns `ErrNextNoChild`, so it’s fine. I guess this isn’t as big a problem, then.

Reply:
> but it returns `ErrNextNoChild`, so it’s fine.

Yes, exactly. Reaching the end would mean that the first call to `NextChild` would fail (as it would normally fail when you skipped all the existing children of the current node). But that shouldn't impact the user's code (you could add a special check for a leaf node so you wouldn't bother to call  `NextChild` even once, but that wouldn't improve the performance much).